tools/clang/lib/Sema/SemaHLSL.cpp

```diff
@@ -8149,22 +8143,29 @@ bool HLSLExternalSource::CanConvert(
       }
     }
 
-    if (const BuiltinType *BT = source->getAs<BuiltinType>()) {
-      BuiltinType::Kind kind = BT->getKind();
-      switch (kind) {
-      case BuiltinType::Kind::UInt:
-      case BuiltinType::Kind::Int:
-      case BuiltinType::Kind::Float:
-      case BuiltinType::Kind::LitFloat:
-      case BuiltinType::Kind::LitInt:
-        if (explicitConversion) {
+    // Handle explicit splats from single element numerical types (scalars, vector1s and matrix1x1s) to aggregate types.
+    if (explicitConversion) {
+      const BuiltinType *sourceSingleElementBuiltinType = source->getAs<BuiltinType>();
+      if (sourceSingleElementBuiltinType == nullptr
+        && hlsl::IsHLSLVecMatType(source)
+        && hlsl::GetElementCount(source) == 1) {
+        sourceSingleElementBuiltinType = hlsl::GetElementTypeOrType(source)->getAs<BuiltinType>();
+      }
+
+      if (sourceSingleElementBuiltinType != nullptr) {
+        BuiltinType::Kind kind = sourceSingleElementBuiltinType->getKind();
+        switch (kind) {
+        case BuiltinType::Kind::UInt:
+        case BuiltinType::Kind::Int:
+        case BuiltinType::Kind::Float:
+        case BuiltinType::Kind::LitFloat:
```

Comment:
Note that this means that 16 or 64-bit types cannot be splat sources, which may be a bit inconsistent?

Reply:
Filed #2047